File: src/main/java/com/lessmarkup/userinterface/model/configuration/NodeAccessModelCollectionManager.java
package com.lessmarkup.userinterface.model.configuration;

import com.google.inject.Inject;
import com.lessmarkup.Constants;
import com.lessmarkup.dataobjects.Node;
import com.lessmarkup.dataobjects.NodeAccess;
import com.lessmarkup.dataobjects.User;
import com.lessmarkup.dataobjects.UserGroup;
import com.lessmarkup.interfaces.cache.EntityChangeType;
import com.lessmarkup.interfaces.data.ChangeTracker;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.data.QueryBuilder;
import com.lessmarkup.interfaces.recordmodel.EditableModelCollection;
import com.lessmarkup.interfaces.structure.NodeAccessType;

import java.util.Collection;
import java.util.LinkedList;
import java.util.List;
import java.util.OptionalLong;

public class NodeAccessModelCollectionManager implements EditableModelCollection<NodeAccessModel> {
    private long nodeId;
    private final DomainModelProvider domainModelProvider;
    private final ChangeTracker changeTracker;

    @Inject
    public NodeAccessModelCollectionManager(DomainModelProvider domainModelProvider, ChangeTracker changeTracker) {
        this.domainModelProvider = domainModelProvider;
        this.changeTracker = changeTracker;
    }

    public void initialize(long nodeId) {
        this.nodeId = nodeId;
    }

    @Override
    public List<Long> readIds(QueryBuilder query, boolean ignoreOrder) {
        return query.from(NodeAccess.class).where("nodeId = $", nodeId).toIdList();
    }

    @Override
    public int getCollectionId() {
        return domainModelProvider.getCollectionId(NodeAccess.class).getAsInt();
    }

    @Override
    public Collection<NodeAccessModel> read(QueryBuilder query, List<Long> ids) {
        List<String> idsString = new LinkedList<>();
        ids.forEach(s -> idsString.add(s.toString()));
        return query.from(NodeAccess.class, "na").where(String.format("na.nodeId = $ AND na." + Constants.DataIdPropertyName() + " IN (%s)", String.join(",", idsString)), nodeId)
            .leftJoin(User.class, "u", "u." + Constants.DataIdPropertyName() + " = na.userId")
            .leftJoin(UserGroup.class, "g", "g." + Constants.DataIdPropertyName() + " = na.groupId")
            .toList(NodeAccessModel.class, "na.accessType, u.email, g.name, na." + Constants.DataIdPropertyName() + " AccessId");
    }

    public boolean isFiltered() { return false; }

    @Override
    public void initialize(OptionalLong objectId, NodeAccessType nodeAccessType) {
    }

    @Override
    public NodeAccessModel createRecord() {
        return new NodeAccessModel();
    }

    @Override
    public void addRecord(NodeAccessModel record) {
        try (DomainModel domainModel = domainModelProvider.createWithTransaction()) {
            NodeAccess access = new NodeAccess();
            access.setAccessType(record.getAccessType());
            access.setNodeId(nodeId);

            if (record.getUser() != null && record.getGroup().length() > 0) {
                long userId = domainModel.query().from(User.class).where("email = $", record.getUser()).first(User.class, "Id").getId();
                access.setUserId(OptionalLong.of(userId));
            }

            if (record.getGroup() != null && record.getGroup().length() > 0) {
                long groupId = domainModel.query().from(UserGroup.class).where("name = $", record.getGroup()).first(UserGroup.class, "Id").getId();
                access.setGroupId(OptionalLong.of(groupId));
            }

            domainModel.create(access);
            changeTracker.addChange(Node.class, nodeId, EntityChangeType.UPDATED, domainModel);
            domainModel.completeTransaction();

            record.setAccessId(access.getId());
        }
    }

    @Override
    public void updateRecord(NodeAccessModel record) {
        try (DomainModel domainModel = domainModelProvider.createWithTransaction()) {
            NodeAccess access = domainModel.query().find(NodeAccess.class, record.getAccessId());
            access.setAccessType(record.getAccessType());

            if (record.getUser() != null && record.getUser().length() > 0) {
                long userId = domainModel.query().from(User.class).where("email = $", record.getUser()).first(User.class).getId();
                access.setUserId(OptionalLong.of(userId));
            }
            else {
                access.setUserId(OptionalLong.empty());
            }

            if (record.getGroup() != null && record.getGroup().length() > 0) {
                long groupId = domainModel.query().from(UserGroup.class).where("name = $", record.getGroup()).first(UserGroup.class).getId();
                access.setGroupId(OptionalLong.of(groupId));
            }
            else {
                access.setGroupId(OptionalLong.empty());
            }

            domainModel.update(access);
            changeTracker.addChange(Node.class, nodeId, EntityChangeType.UPDATED, domainModel);
            domainModel.completeTransaction();
        }
    }

    @Override
    public boolean deleteRecords(Collection<Long> recordIds) {
        try (DomainModel domainModel = domainModelProvider.createWithTransaction()) {
            boolean hasChanges = false;

            for (long recordId : recordIds) {
                domainModel.delete(NodeAccess.class, recordId);
                hasChanges = true;
            }

            if (hasChanges) {
                changeTracker.addChange(Node.class, nodeId, EntityChangeType.UPDATED, domainModel);
                domainModel.completeTransaction();
            }

            return hasChanges;
        }
    }
}

File: src/main/scala/com/lessmarkup/userinterface/model/configuration/NodeAccessModelCollectionManager.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.model.configuration

import com.google.inject.Inject
import com.lessmarkup.Constants
import com.lessmarkup.dataobjects.{Node, NodeAccess, User, UserGroup}
import com.lessmarkup.interfaces.annotations.UseInstanceFactory
import com.lessmarkup.interfaces.cache.{EntityChangeType, InstanceFactory}
import com.lessmarkup.interfaces.data.{ChangeTracker, DomainModel, DomainModelProvider, QueryBuilder}
import com.lessmarkup.interfaces.recordmodel.EditableModelCollection

class NodeAccessModelCollectionManagerFactory @Inject() (domainModelProvider: DomainModelProvider, changeTracker: ChangeTracker)
  extends InstanceFactory {
  override def createInstance(params: Any*): AnyRef = {
    new NodeAccessModelCollectionManager(domainModelProvider, changeTracker, params.head.asInstanceOf)
  }
}

@UseInstanceFactory(classOf[NodeAccessModelCollectionManagerFactory])
class NodeAccessModelCollectionManager @Inject() (
  domainModelProvider: DomainModelProvider,
  changeTracker: ChangeTracker,
  nodeId: Long)
  extends EditableModelCollection[NodeAccessModel] {

  def readIds(query: QueryBuilder, ignoreOrder: Boolean): Seq[Long] = {
    query.from(classOf[NodeAccess])
      .where("nodeId = $", nodeId).toIdList
  }

  def getCollectionId: Int = {
    domainModelProvider.getCollectionId(classOf[NodeAccess]).get
  }

  def read(query: QueryBuilder, ids: Seq[Long]): Seq[NodeAccessModel] = {
    query.from(classOf[NodeAccess], Option("na"))
      .where(s"na.nodeId = $$ AND na.${Constants.DataIdPropertyName} IN (${ids.map(_.toString).mkString(",")}})", nodeId)
      .leftJoin(classOf[User], "u", "u." + Constants.DataIdPropertyName + " = na.userId")
      .leftJoin(classOf[UserGroup], "g", "g." + Constants.DataIdPropertyName + " = na.groupId")
      .toList(classOf[NodeAccessModel], Option("na.accessType, u.email, g.name, na." + Constants.DataIdPropertyName + " AccessId"))
  }

  def isFiltered: Boolean = false

  def createRecord: NodeAccessModel = {
    new NodeAccessModel
  }

  def addRecord(record: NodeAccessModel) {
    val domainModel: DomainModel = domainModelProvider.createWithTransaction
    try {

      val access = new NodeAccess
      access.accessType = record.accessType
      access.nodeId = nodeId

      if (record.user != null && record.group.length > 0) {
        val userId: Long = domainModel.query
          .from(classOf[User])
          .where("email = $", record.user)
          .first(classOf[User], Option("Id"))
          .get.id
        access.userId = Option(userId)
      }

      if (record.group != null && record.group.length > 0) {
        val groupId: Long = domainModel.query
          .from(classOf[UserGroup])
          .where("name = $", record.group)
          .first(classOf[UserGroup], Option("Id"))
          .get.id
        access.groupId = Option(groupId)
      }
      domainModel.create(access)
      changeTracker.addChange(classOf[Node], nodeId, EntityChangeType.UPDATED, domainModel)
      domainModel.completeTransaction()
      record.accessId = access.id
    } finally {
      if (domainModel != null) domainModel.close()
    }
  }

  def updateRecord(record: NodeAccessModel) {
    val domainModel: DomainModel = domainModelProvider.createWithTransaction
    try {
      val access: NodeAccess = domainModel.query.find(classOf[NodeAccess], record.accessId).get
      access.accessType = record.accessType
      if (record.user != null && record.user.length > 0) {
        val userId: Long = domainModel.query
          .from(classOf[User])
          .where("email = $", record.user)
          .first(classOf[User], None)
          .get.id
        access.userId = Option(userId)
      }
      else {
        access.userId = None
      }
      if (record.group != null && record.group.length > 0) {
        val groupId: Long = domainModel.query
          .from(classOf[UserGroup])
          .where("name = $", record.group)
          .first(classOf[UserGroup], None)
          .get.id
        access.groupId = Option(groupId)
      }
      else {
        access.groupId = None
      }
      domainModel.update(access)
      changeTracker.addChange(classOf[Node], nodeId, EntityChangeType.UPDATED, domainModel)
      domainModel.completeTransaction()
    } finally {
      if (domainModel != null) domainModel.close()
    }
  }

  def deleteRecords(recordIds: Seq[Long]): Boolean = {
    val domainModel: DomainModel = domainModelProvider.createWithTransaction
    try {
      var hasChanges: Boolean = false
      for (recordId <- recordIds) {
        domainModel.delete(classOf[NodeAccess], recordId)
        hasChanges = true
      }
      if (hasChanges) {
        changeTracker.addChange(classOf[Node], nodeId, EntityChangeType.UPDATED, domainModel)
        domainModel.completeTransaction()
      }
      hasChanges
    } finally {
      if (domainModel != null) domainModel.close()
    }
  }
}
